packages/protocol/contracts/common/FeeHandler.sol

```diff
@@ -61,39 +89,153 @@ contract FeeHandler is Ownable, Initializable, UsingRegistry, ICeloVersionedCont
+    FixidityLib.Fraction memory fraction = FixidityLib.wrap(newFraction);
+    require(
+      FixidityLib.lte(fraction, FixidityLib.fixed1()),
+      "Burn fraction must be less than or equal to 1"
+    );
+    burnFraction = fraction;
+    emit BurnFractionSet(newFraction);
+  }
+
+  function setBurnFraction(uint256 fraction) external onlyOwner {
+    return _setBurnFraction(fraction);
+  }
+
+  function sell(address tokenAddress) external {
+    return _sell(tokenAddress);
+  }
+
+  function _addToken(address tokenAddress, address handlerAddress) private {
+    IFeeHandlerSeller(handlerAddress);
+
+    // Check that the contract implements the interface
+    TokenState storage tokenState = tokenStates[tokenAddress];
+    tokenState.active = true;
+    tokenState.handler = handlerAddress;
+
+    activeTokens.push(tokenAddress);
+  }
+
+  function addToken(address tokenAddress, address handlerAddress) external onlyOwner {
+    _addToken(tokenAddress, handlerAddress);
+  }
+
+  function getActiveTokens() public view returns (address[] memory) {
+    return activeTokens;
+  }
+
+  function removeToken(address tokenAddress) external {}
+
+  function _sell(address tokenAddress) private onlyWhenNotFrozen nonReentrant {
+    IERC20 token = IERC20(tokenAddress);
+
+    TokenState storage tokenState = tokenStates[tokenAddress];
+    FixidityLib.Fraction memory balanceOfTokenToBurn = FixidityLib.newFixed(
+      token.balanceOf(address(this)).sub(tokenState.toDistribute)
+    );
+
+    uint256 balanceToBurn = (burnFraction.multiply(balanceOfTokenToBurn).fromFixed());
+
+    tokenState.toDistribute += (token.balanceOf(address(this)).sub(balanceToBurn));
+
+    // small numbers cause rounding errors and zero case should be skipped
+    if (balanceToBurn <= MIN_BURN) {
+      return;
+    }
+
+    if (dailyBurnLimitHit(tokenAddress, balanceToBurn)) {
+      // in case the limit is hit, burn the max possible
+      // TODO move to state
+      balanceToBurn = tokenState.currentDayLimit;
+      emit DailyLimitHit(tokenAddress, balanceToBurn);
+    }
+
+    token.transfer(address(tokenState.handler), balanceToBurn);
+    IFeeHandlerSeller handler = IFeeHandlerSeller(tokenState.handler);
+    handler.sell(
+      tokenAddress,
+      registry.getAddressForOrDie(GOLD_TOKEN_REGISTRY_ID),
+      balanceToBurn,
+      FixidityLib.unwrap(tokenState.maxSlippage)
+    );
+
+    tokenState.pastBurn = tokenState.pastBurn.add(balanceToBurn);
+    updateLimits(tokenAddress, balanceToBurn);
+
+    emit SoldAndBurnedToken(tokenAddress, balanceToBurn);
+  }
+
+  function distribute(address tokenAddress) external {
+    return _distribute(tokenAddress);
+  }
+
+  function _distribute(address tokenAddress) private {
+    require(feeBeneficiary != address(0), "Can't distribute to the zero address");
+    TokenState storage tokenState = tokenStates[tokenAddress];
+    IERC20 token = IERC20(tokenAddress);
+    token.transfer(feeBeneficiary, tokenState.toDistribute);
```

Comment:
What if contract doesnt have enought of ERC20 token? We could instead distribute whatever contract has and then lower toDistribute